src/Expr.hs

```diff
@@ -0,0 +1,56 @@
+module Expr where
+
+data Expr = Num Double
+            | Var String
+            | Plus Expr Expr
+            | Minus Expr Expr
+            | Mult Expr Expr
+            | Div Expr Expr
+            | Pow Expr Expr
+            | Abs Expr
+            | UnaryMinus Expr
+
+instance Eq Expr where
+  (==) = exprEquals
+
+exprEquals :: Expr -> Expr -> Bool
+exprEquals (Num a) (Num b) = a == b
+exprEquals (Var a) (Var b) = a == b
+exprEquals (Plus a1 b1) (Plus a2 b2) = exprEquals a1 a2 && exprEquals b1 b2 || exprEquals a1 b2 && exprEquals b1 a2
+exprEquals (Minus a1 b1) (Minus a2 b2) = exprEquals a1 a2 && exprEquals b1 b2
+exprEquals (Mult a1 b1) (Mult a2 b2) = exprEquals a1 a2 && exprEquals b1 b2 || exprEquals a1 b2 && exprEquals b1 a2
+exprEquals (Div a1 b1) (Div a2 b2) = exprEquals a1 a2 && exprEquals b1 b2
+exprEquals (Pow a1 b1) (Pow a2 b2) = exprEquals a1 a2 && exprEquals b1 b2
+exprEquals (Abs a1) (Abs a2) = exprEquals a1 a2
+exprEquals (UnaryMinus (Num a)) (Num b) = a == -b
+exprEquals (Num a) (UnaryMinus (Num b)) = a == -b
+exprEquals (UnaryMinus a1) (UnaryMinus a2) = exprEquals a1 a2
+exprEquals _ _ = False
```

Comment:
Eq should implement syntactic equality. Any reasons why you treat unary negation differently?

Reply:
Details like unary negation are lost when converting to String and parsing back, so these equations should be considered equal